File: quicksort.py
import random

def quicksort(arr, low, high):
    if low < high:
        pivot = partition(arr, low, high)
        quicksort(arr, low, pivot - 1)
        quicksort(arr, pivot + 1, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low
    for j in range(low, high):
        if arr[j] <= pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[high] = arr[high], arr[i]
    return i

import time

nums = [random.randint(0, 9999) for _ in range(10000)]
print("Before:", nums[:10])
start = time.perf_counter()
quicksort(nums, 0, len(nums) - 1)
elapsed = time.perf_counter() - start
print("After: ", nums[:10])
print(f"Sorted 10000 numbers in {elapsed*1000:.3f} ms")

File: Quicksort.java
import java.util.Random;

public class Quicksort {
    static void quicksort(int[] arr, int low, int high) {
        if (low < high) {
            int pivot = partition(arr, low, high);
            quicksort(arr, low, pivot - 1);
            quicksort(arr, pivot + 1, high);
        }
    }

    static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low;
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                int tmp = arr[i];
                arr[i] = arr[j];
                arr[j] = tmp;
                i++;
            }
        }
        int tmp = arr[i];
        arr[i] = arr[high];
        arr[high] = tmp;
        return i;
    }

    public static void main(String[] args) {
        Random rng = new Random(42);
        int[] nums = new int[10000];
        for (int i = 0; i < nums.length; i++) {
            nums[i] = rng.nextInt(10000);
        }
        System.out.print("Before: ");
        for (int i = 0; i < 10; i++) { System.out.print(nums[i] + " "); }
        System.out.println();

        long start = System.nanoTime();
        quicksort(nums, 0, nums.length - 1);
        long elapsed = System.nanoTime() - start;

        System.out.print("After:  ");
        for (int i = 0; i < 10; i++) { System.out.print(nums[i] + " "); }
        System.out.println();
        System.out.printf("Sorted 10000 numbers in %.3f ms%n", elapsed / 1_000_000.0);
    }
}
